org.eclipse.xtext/src/org/eclipse/xtext/generator/trace/AbstractTraceRegionToString.java

```diff
@@ -0,0 +1,434 @@
+	}
+
+	protected String render(AbstractTraceRegionToString.LocationHandle loc) {
+		String prefix = loc.id >= 0 ? Integer.toString(loc.id, this.radix) + ": " : "";
+		String name = loc.location.getClass().getSimpleName();
+		String path = loc.location.getSrcRelativePath() == null ? "" : ("," + loc.location.getSrcRelativePath());
+		return prefix + name + "[" + loc.location.getOffset() + "," + loc.location.getLength() + path + "]";
+	}
+
+	protected void render(AbstractTraceRegionToString.RegionHandle region, int idW, int offsetW, int lengthW,
+			int indent, List<String> result) {
+		String id = Strings.padStart(Integer.toString(region.id, this.radix), idW, '0');
+		String debug = region.region.isUseForDebugging() ? "D" : " ";
+		String offset = Strings.padStart(Integer.toString(region.region.getMyOffset()), offsetW, '0');
+		String length = Strings.padStart(Integer.toString(region.region.getMyLength()), lengthW, '0');
+		String space = Strings.repeat(" ", indent);
+		String name = region.region.getClass().getSimpleName();
+		String locations = join(map(region.locations, it -> render(it)), ", ");
+		String loc = debug + " " + offset + "-" + length + space;
+		String header = id + ": " + loc + name + " -> " + locations;
+		if (region.children.isEmpty()) {
+			result.add(header);
+		} else {
+			result.add((header + " {"));
+			for (RegionHandle child : region.children) {
+				render(child, idW, offsetW, lengthW, (indent + 2), result);
+			}
+			result.add((id + ": " + Strings.repeat(" ", loc.length())) + "}");
+		}
+	}
+
+	protected String render() {
+		File localFile = new File(null);
+		LinkedHashMap<SourceRelativeURI, File> remoteFiles = CollectionLiterals.newLinkedHashMap();
+		List<RegionHandle> roothandles = CollectionLiterals.newArrayList();
+		int maxid = collect(getTrace(), 1, localFile, remoteFiles, roothandles);
+		int idwidth = Integer.toString(maxid, radix).length();
+		addAll(localFile.lines, render(localFile, idwidth));
+		for (File file : remoteFiles.values()) {
+			addAll(file.lines, render(file, idwidth));
+		}
+		int localWidth = Math.max(max(map(localFile.lines, (String it) -> Integer.valueOf(it.length()))),
+				getLocalTitle().length() + 2);
+		int remoteWidth = max(map(remoteFiles.values(),
+				(File it) -> Math.max(max(map(it.lines, (String it_1) -> Integer.valueOf(it_1.length()))),
+						getRemoteTitle(it.uri).length() + 2)));
+		localFile.lines.add(0, title(null, localWidth));
+		for (File file_1 : remoteFiles.values()) {
+			file_1.lines.add(0, title(file_1.uri, remoteWidth));
+		}
+		List<String> left = localFile.lines;
+		List<String> right = toList(concat(map(remoteFiles.values(), it -> it.lines)));
+		ArrayList<String> result = CollectionLiterals.newArrayList();
+		if (showLegend) {
+			result.add(
+					"Regions are surrounded by [N[ ... ]N]. Regions on the left and right with the same N are associated.");
+		}
+		for (int i = 0; ((i < ((Object[]) Conversions.unwrapArray(left, Object.class)).length)
```

Comment:
`Conversions.unwrapArray` this really isnt needed

Reply:
Fixed, but please have a look if you agree that I got the meaning right.